// @textile/powergate-client/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPow = exports.ListSelect = exports.powTypes = exports.adminTypes = void 0;
const user_pb_1 = require("@textile/grpc-powergate-client/dist/powergate/user/v1/user_pb");
const user_pb_service_1 = require("@textile/grpc-powergate-client/dist/powergate/user/v1/user_pb_service");
const admin_1 = require("./admin");
const data_1 = require("./data");
const deals_1 = require("./deals");
const storage_config_1 = require("./storage-config");
const storage_info_1 = require("./storage-info");
const storage_jobs_1 = require("./storage-jobs");
Object.defineProperty(exports, "ListSelect", { enumerable: true, get: function () { return storage_jobs_1.ListSelect; } });
const util_1 = require("./util");
const wallet_1 = require("./wallet");
exports.adminTypes = __importStar(require("@textile/grpc-powergate-client/dist/powergate/admin/v1/admin_pb"));
exports.powTypes = __importStar(require("@textile/grpc-powergate-client/dist/powergate/user/v1/user_pb"));
const defaultConfig = {
    host: util_1.host,
    transport: util_1.getTransport(),
};
/**
 * Creates a new Powergate client
 * @param config A config object that changes the behavior of the client
 * @returns A Powergate client API
 */
const createPow = (config) => {
    const c = Object.assign(Object.assign({}, defaultConfig), removeEmpty(config));
    const { getMeta, getHeaders, setToken, setAdminToken } = util_1.useTokens(c.authToken, c.adminToken);
    const client = new user_pb_service_1.UserServiceClient(c.host, c);
    return {
        host: c.host,
        setToken,
        setAdminToken,
        buildInfo: () => util_1.promise((cb) => client.buildInfo(new user_pb_1.BuildInfoRequest(), cb), (resp) => resp.toObject()),
        userId: () => util_1.promise((cb) => client.userIdentifier(new user_pb_1.UserIdentifierRequest(), getMeta(), cb), (res) => res.toObject()),
        storageConfig: storage_config_1.createStorageConfig(c, getMeta),
        data: data_1.createData(c, getMeta, getHeaders),
        wallet: wallet_1.createWallet(c, getMeta),
        deals: deals_1.createDeals(c, getMeta),
        storageInfo: storage_info_1.createStorageInfo(c, getMeta),
        storageJobs: storage_jobs_1.createStorageJobs(c, getMeta),
        admin: admin_1.createAdmin(c, getMeta),
    };
};
exports.createPow = createPow;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const removeEmpty = (obj) => {
    Object.keys(obj).forEach((key) => {
        if (obj[key] && typeof obj[key] === "object")
            removeEmpty(obj[key]);
        else if (obj[key] === undefined)
            delete obj[key];
    });
    return obj;
};
